src/net_processing.cpp

```diff
@@ -949,6 +949,16 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                 BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
                 if (mi != mapBlockIndex.end())
                 {
+                    if (mi->second->nChainTx && !mi->second->IsValid(BLOCK_VALID_SCRIPTS) &&
```

Comment:
commented on IRC, but yes i think we need a fix here.

Otherwise we can be in a situation where we stall a peer who asks for a block we just announced to them.

i.e. We try and connect A and A' near simultaneously as new tips.  We announce both, but whichever gets connected first, the other we will stall requests for.  This can happen even without multi-threaded ProcessMessages via a race with submitblock.